96.md

```diff
@@ -0,0 +1,221 @@
+    "free": {
+      "name": "Free Tier",
+      "url": "https://...", // payment etc
+      "max_byte_size": 10485760,
+      // Range in days / 0 for no expiration
+      // [7, 0] means it may vary from 7 days to unlimited persistence,
+      // [0, 0] means it has no expiration
+      // early expiration may be due to low traffic or any other factor
+      "file_expiration": [14, 90],
+      "image_transformations": ["resizing"]
+    }
+  }
+}
+```
+
+Note that if using a CDN, the `server` MAY issue a 302 redirect from
+GET requests to the `api_url` or `download_url` route to the CDN url instead
+of assigning the CDN directly as the `download_url` route.
+This is because some NIP-96 integrations
+MAY require NIP-98 authentication for downloads,
+which the CDN won't be able to support.
+
+### Relay Hints
+
+A nostr relay MAY advertise someone else's HTTP file storage server by
+adding a `/.well-known/nostr/nip96.json` with "recommended_url" field
+pointing to the url where the server hosts its own
+`/.well-known/nostr/nip96.json`. In this case, the "api_url" must
+be an empty string.
+
+If the nostr relay is also an HTTP file storage server,
+it must use the "api_url" field instead.
+
+### List of Supporting File Storage Servers
+
+| Name | Domain |
+| ---- | ------ |
+| -    | -      |
+
+## Upload
+
+A file can be uploaded one at a time to `https://your-file-server.example/custom-api-path` (route from `https://your-file-server.example/.well-known/nostr/nip96.json` "api_url" field) as `multipart/form-data` content type using `POST` method with the file object set to the `file` form data field.
+
+`Clients` must add a [NIP-98](98.md) `Authorization` header with the encoded `payload` tag set to the base64-encoded 256-bit SHA-256 hash of the file (not the hash of the whole request body).
+If using an html form, use an `Authorization` form data field instead.
+
+These following **optional** form data fields MAY be used by `servers` and SHOULD be sent by `clients`:
+- `expiration`: string of the UNIX timestamp in seconds. Empty string if file should be stored forever. The server isn't required to honor this;
+- `size`: string of the file byte size. This is just a value the server can use to reject early if the file size exceeds the server limits;
+- `alt`: strict description text for visibility-impaired users;
+- `caption`: loose description;
+- `content_type`: mime type such as "image/jpeg".
+
+Others custom form data fields may be used depending on specific `server` support.
+The `server` isn't required to store any metadata sent by `clients`.
+
+Note for `clients`: if using an HTML form, it is important for the `file` form field to be the **last** one, or be re-ordered right before sending or be appended as the last field of XHR2's FormData object.
+
+The `filename` embedded in the file may not be honored by the `server`, which could internally store just the SHA-256 hash value as the file name, ignoring extra metadata.
```

Comment:
This will be possible with the x tag appended to the url (with NIP-54) or with the x tag (with NIP-94).
We can't use the filename because other file urls not using NIP-96 most times won't use the convention of having the SHA-256 as filename.

Reply:
> SHA-256 of the file content (before any alterations) seems to be the good starting point.

In practice that's what this PR wants. A client must be able to download at `https://download.route/path/<original-file-hash>`. In the future when the file gets deleted, a client can try another server's download route appending the original file's hash as if it was the filename.